database/postgres/postgres.go

```diff
@@ -167,13 +169,53 @@ func (p *Postgres) Run(migration io.Reader) error {
 	// run migration
 	query := string(migr[:])
 	if _, err := p.conn.ExecContext(context.Background(), query); err != nil {
-		// TODO: cast to postgress error and get line number
+		if pgErr, ok := err.(*pq.Error); ok {
+			var line uint
+			var col uint
+			var lineColOK bool
+			if pgErr.Position != "" {
+				if pos, err := strconv.ParseUint(pgErr.Position, 10, 64); err == nil {
+					if line, col, ok = computeLineFromPos(query, uint(pos)); ok {
+						lineColOK = true
+					}
+				}
+			}
+			message := fmt.Sprintf("migration failed: %s", pgErr.Message)
+			if lineColOK {
+				message = fmt.Sprintf("%s (column %d)", message, col)
+			}
+			if pgErr.Detail != "" {
+				message = fmt.Sprintf("%s, %s", message, pgErr.Detail)
+			}
+			return database.Error{OrigErr: err, Err: message, Query: migr, Line: line}
+		}
 		return database.Error{OrigErr: err, Err: "migration failed", Query: migr}
 	}
 
 	return nil
 }
 
+func computeLineFromPos(s string, pos uint) (uint, uint, bool) {
+	newLine := "\n"
+	if i := strings.Index(s, "\r\n"); i >= 0 {
+		newLine = "\r\n"
+	}
+	lines := strings.Split(s, newLine)
```

Comment:
You should be able to avoid splitting strings and instead count the number of newlines.
e.g.
```
lineNum := strings.Count(s[:pos], newline)
colNum := pos - strings.LastIndex(s[:pos], newline)
```

Reply:
Thanks for the tip! I updated the function using this approach, but also improved it by supporting non-ascii characters (tests are updated as well).